jl4-lsp/src/LSP/L4/SemanticTokens.hs

```diff
@@ -214,3 +236,98 @@ instance ToSemTokens () PosToken PosToken where
   toSemTokens t = do
     ctx <- ask
     pure $ maybe [] toList (fromSemanticTokenContext ctx t)
+
+--- ---------------------------------------------------------------------------
+-- Semantic Tokens for typechecked AST
+--- ---------------------------------------------------------------------------
+
+instance ToSemTokens EntityInfo PosToken Resolved where
+  toSemTokens r = do
+    entityInfo <- asks (.semanticTokenContext)
+    case r of
+      Def _ n -> modifier entityInfo $ withModifier defVar $ toSemTokens n
+      Ref orig _ _ -> modifier entityInfo $ toSemTokens orig
+      OutOfScope _ n -> modifier entityInfo $ toSemTokens n
+    where
+      uniq = getUnique r
+      modifier entityInfo =
+        case Map.lookup uniq entityInfo of
+          -- Strictly speaking, this is an internal invariant violation.
+          -- However, we likely shouldn't care during syntax highlighting.
+          Nothing -> id
```

Comment:
i think we should put an `assert` on those cases.

Reply:
This comment is out of date, many `Resolved` are not part of the `entityInfo`, for example `type variables`.